File: backend/internal/social/content/entity/comment.go
package entity

import (
	"fmt"
	"time"

	"github.com/google/uuid"
)

// CommentType represents the type of comment.
type CommentType string

const (
	// CommentTypeNormal is a regular text comment.
	CommentTypeNormal CommentType = "normal"
	// CommentTypeCommerceReference is a comment with an attached commerce reference.
	CommentTypeCommerceReference CommentType = "commerce_reference"
)

// IsValid checks if the comment type is valid.
func (t CommentType) IsValid() bool {
	switch t {
	case CommentTypeNormal, CommentTypeCommerceReference:
		return true
	default:
		return false
	}
}

// CommentTargetType represents the target entity type for generic comments.
//
// ZERO LEGACY MODE: Only 'content' is canonical for V1
// Comments are only for content (posts and requests)
type CommentTargetType string

const (
	// TargetContent targets a content post (canonical V1).
	TargetContent CommentTargetType = "content"
)

// IsValid checks if the target type is valid.
func (t CommentTargetType) IsValid() bool {
	return t == TargetContent
}

// Comment represents a comment on a target entity.
//
// ZERO LEGACY MODE: Only content target is canonical for V1
//
// SHARE ALIGNMENT V1: Commerce reference comments use ShareReference
// - type=commerce_reference must have reference set
// - reference.targetType = for_sale | auction
// - reference.targetId = canonical commerce resource ID
// - reference.preview = cached display data
type Comment struct {
	ID         uuid.UUID
	AuthorID   uuid.UUID
	Body       *string
	Type       CommentType
	Reference  *ShareReference   // Share reference for commerce_reference comments (fixed-price sale or auction)
	TargetID   uuid.UUID         // Target entity ID (corresponds to TargetType)
	TargetType CommentTargetType // Target entity type
	ParentID   *uuid.UUID        // Parent comment ID for threaded replies (null for top-level)
	CreatedAt  time.Time
	UpdatedAt  time.Time
	DeletedAt  *time.Time
}

// ErrInvalidComment is returned when comment validation fails.
type ErrInvalidComment struct {
	Reason string
}

func (e *ErrInvalidComment) Error() string {
	return fmt.Sprintf("invalid comment: %s", e.Reason)
}

// ErrCommerceReferenceOnPost is returned when trying to add a commerce reference to non-request content.
type ErrCommerceReferenceOnPost struct {
	Reason string
}

func (e *ErrCommerceReferenceOnPost) Error() string {
	return fmt.Sprintf("commerce reference only allowed on request-type content: %s", e.Reason)
}

// NewComment creates a new normal comment.
// PRECONDITION: body must not be empty.
func NewComment(targetID, authorID uuid.UUID, body string) (*Comment, error) {
	if body == "" {
		return nil, &ErrInvalidComment{Reason: "body cannot be empty for normal comments"}
	}

	now := time.Now()
	return &Comment{
		ID:         uuid.New(),
		AuthorID:   authorID,
		Body:       &body,
		Type:       CommentTypeNormal,
		TargetID:   targetID,
		TargetType: TargetContent,
		CreatedAt:  now,
		UpdatedAt:  now,
	}, nil
}

// NewCommerceReferenceComment creates a new commerce reference comment.
// Body is optional (can be empty for commerce-reference-only comments).
//
// SHARE ALIGNMENT V1: Accepts ShareReference instead of forSaleID
// - reference must have targetType = for_sale or auction
// - reference must have valid targetId
// - reference.preview contains cached display data
func NewCommerceReferenceComment(targetID, authorID uuid.UUID, reference *ShareReference, body *string) (*Comment, error) {
	now := time.Now()
	return &Comment{
		ID:         uuid.New(),
		AuthorID:   authorID,
		Body:       body,
		Type:       CommentTypeCommerceReference,
		Reference:  reference,
		TargetID:   targetID,
		TargetType: TargetContent,
		CreatedAt:  now,
		UpdatedAt:  now,
	}, nil
}

// Validate checks if the comment is valid according to its type.
func (c *Comment) Validate() error {
	if !c.Type.IsValid() {
		return &ErrInvalidComment{Reason: fmt.Sprintf("invalid comment type: %s", c.Type)}
	}

	switch c.Type {
	case CommentTypeNormal:
		if c.Body == nil || *c.Body == "" {
			return &ErrInvalidComment{Reason: "body cannot be empty for normal comments"}
		}
	case CommentTypeCommerceReference:
		if c.Reference == nil || !c.Reference.IsValid() || !c.Reference.TargetType.IsValid() {
			return &ErrInvalidComment{Reason: "reference is required for commerce reference comments"}
		}
	}

	return nil
}

// IsNormal returns true if this is a normal text comment.
func (c *Comment) IsNormal() bool {
	return c.Type == CommentTypeNormal
}

// IsCommerceReference returns true if this is a commerce reference comment.
func (c *Comment) IsCommerceReference() bool {
	return c.Type == CommentTypeCommerceReference
}

// IsDeleted returns true if the comment has been soft deleted.
func (c *Comment) IsDeleted() bool {
	return c.DeletedAt != nil
}

// HasParent returns true if this is a reply to another comment.
func (c *Comment) HasParent() bool {
	return c.ParentID != nil
}

File: apps/mobile/lib/domains/social/comment/domain/entities/comment.dart
import 'package:equatable/equatable.dart';
import 'package:labuda/shared/attachment/entities/share_reference.dart';
import 'package:labuda/shared/governance/content_lifecycle.dart';

/// Comment Entity - canonical domain entity for comments system.
///
/// Comment is a social interaction object, NOT a commerce object.
/// reference represents seller response/recommendation ONLY.
/// Does NOT represent binding offers, negotiations, or transactions.
/// Reply max depth = 1 (top-level comments can be replied, replies cannot).
///
/// SINGLE REFERENCE MODEL:
/// - reference uses ShareReference (canonical cross-domain reference)
/// - targetType = "for_sale" / "auction" / "content"
///
/// Pure domain entity - no Flutter/Firebase dependencies.
class Comment extends Equatable {
  final String id;
  final String authorId;
  final String contentId;
  // Author info embedded in response for proper UI rendering
  final String authorUsername;
  final String? authorAvatarUrl;
  final String? body;
  final String type;
  final ShareReference? reference; // Unified reference using ShareReference
  final String? parentId; // Set for replies (max depth = 1)
  final DateTime createdAt;
  final DateTime? updatedAt;
  final DateTime? deletedAt;

  /// E3.1 — Canonical governance lifecycle for the comment AUTHOR identity
  /// (PublicCard.UserCard.Lifecycle on the comment-author seam).
  ///
  /// Independent of [deletedAt] (which is the COMMENT's own soft-delete
  /// state, not the author's). Defaults to active for null / unknown /
  /// missing wire values so legacy backend payloads keep rendering
  /// today's behavior.
  ///
  /// Mobile preparation only — backend currently emits `null` for this
  /// field. The mobile redaction seam is wired ahead of any backend
  /// activation per E3.1 doctrine.
  final ContentLifecycle authorLifecycle;

  const Comment({
    required this.id,
    required this.authorId,
    required this.contentId,
    required this.authorUsername,
    this.authorAvatarUrl,
    this.body,
    required this.type,
    this.reference,
    this.parentId,
    required this.createdAt,
    this.updatedAt,
    this.deletedAt,
    this.authorLifecycle = ContentLifecycle.active,
  });

  /// Check if this is a commerce reference comment (seller response)
  bool get isCommerceReference => type == 'commerce_reference';

  /// Check if this comment has a reference attached
  bool get hasReference => reference != null;

  /// Check if this is a normal comment
  bool get isNormal => type == 'normal';

  /// Check if this comment has been deleted (soft delete)
  bool get isDeleted => deletedAt != null;

  /// Check if this is a reply (has parent)
  bool get isReply => parentId != null;

  /// Check if this is a top-level comment (can be replied to)
  bool get isTopLevel => parentId == null;

  @override
  List<Object?> get props => [
    id,
    authorId,
    contentId,
    authorUsername,
    authorAvatarUrl,
    body,
    type,
    reference,
    parentId,
    createdAt,
    updatedAt,
    deletedAt,
    authorLifecycle,
  ];

  Comment copyWith({
    String? id,
    String? authorId,
    String? contentId,
    String? authorUsername,
    String? authorAvatarUrl,
    String? body,
    String? type,
    ShareReference? reference,
    String? parentId,
    DateTime? createdAt,
    DateTime? updatedAt,
    DateTime? deletedAt,
    ContentLifecycle? authorLifecycle,
  }) {
    return Comment(
      id: id ?? this.id,
      authorId: authorId ?? this.authorId,
      contentId: contentId ?? this.contentId,
      authorUsername: authorUsername ?? this.authorUsername,
      authorAvatarUrl: authorAvatarUrl ?? this.authorAvatarUrl,
      body: body ?? this.body,
      type: type ?? this.type,
      reference: reference ?? this.reference,
      parentId: parentId ?? this.parentId,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      deletedAt: deletedAt ?? this.deletedAt,
      authorLifecycle: authorLifecycle ?? this.authorLifecycle,
    );
  }
}

/// Cursor page of comments returned by the canonical list endpoint.
///
/// C-IDENT / C-CURSOR:
/// - [comments] in backend ASC (oldest-first) order.
/// - [nextCursor] is the opaque cursor to pass to the next page fetch; null
///   means the list is exhausted (the backend list carries no `total`).
class CommentPage extends Equatable {
  final List<Comment> comments;
  final String? nextCursor;

  const CommentPage({required this.comments, this.nextCursor});

  @override
  List<Object?> get props => [comments, nextCursor];
}

/// Comment type enum matching backend CommentType
///
/// Aligned with backend/internal/domain/content/entity/comment.go
enum CommentType {
  /// Normal text comment
  normal('normal'),

  /// Commerce reference comment (seller response with resource attachment)
  commerceReference('commerce_reference');

  const CommentType(this.value);
  final String value;

  static CommentType fromString(String value) {
    return CommentType.values.firstWhere(
      (type) => type.value == value,
      orElse: () => CommentType.normal,
    );
  }
}

/// Comment target type enum matching backend CommentTargetType.
///
/// Only 'content' target is canonical for V1.
enum CommentTargetType {
  /// Content — the only canonical target for V1.
  content('content');

  const CommentTargetType(this.value);
  final String value;

  static CommentTargetType fromString(String value) {
    return CommentTargetType.values.firstWhere(
      (type) => type.value == value,
      orElse: () => CommentTargetType.content,
    );
  }
}
